File: src/lib/walk.js
import fs from "node:fs";
import path from "node:path";

const DEFAULT_IGNORE = new Set([
  "node_modules",
  "target",
  "build",
  "dist",
  "out",
  ".git",
  ".idea",
  ".vscode",
  ".mvn",
  "frontend-generated",
]);

/**
 * Recursively walk `root`, yielding absolute file paths whose basename matches
 * one of the provided extensions. Directories in the ignore set are skipped.
 *
 * @param {string} root
 * @param {object} [options]
 * @param {string[]} [options.extensions] e.g. [".java", ".css"]
 * @param {Set<string>} [options.ignore]
 * @returns {string[]}
 */
export function walk(root, { extensions, ignore = DEFAULT_IGNORE } = {}) {
  const results = [];
  const exts = extensions ? new Set(extensions.map((e) => e.toLowerCase())) : null;

  const stack = [root];
  while (stack.length > 0) {
    const dir = stack.pop();
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      continue; // unreadable directory — skip
    }
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!ignore.has(entry.name)) stack.push(full);
      } else if (entry.isFile()) {
        if (!exts || exts.has(path.extname(entry.name).toLowerCase())) {
          results.push(full);
        }
      }
    }
  }
  return results.sort();
}

File: go/internal/lib/walk.go
package lib

import (
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// defaultIgnore lists directory names that are never descended into.
// Mirrors DEFAULT_IGNORE in src/lib/walk.js.
var defaultIgnore = map[string]bool{
	"node_modules":       true,
	"target":             true,
	"build":              true,
	"dist":               true,
	"out":                true,
	".git":               true,
	".idea":              true,
	".vscode":            true,
	".mvn":               true,
	"frontend-generated": true,
}

// Walk returns absolute paths of regular files under root whose extension is in
// exts (compared case-insensitively, including the leading dot). An empty exts
// matches every file. Unreadable directories are skipped rather than failing the
// whole walk. Results are sorted for deterministic output.
//
// Mirrors walk() in src/lib/walk.js.
func Walk(root string, exts []string) []string {
	extSet := make(map[string]bool, len(exts))
	for _, e := range exts {
		extSet[strings.ToLower(e)] = true
	}

	var results []string
	stack := []string{root}
	for len(stack) > 0 {
		dir := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		entries, err := os.ReadDir(dir)
		if err != nil {
			continue // unreadable directory — skip
		}
		for _, entry := range entries {
			full := filepath.Join(dir, entry.Name())
			if entry.IsDir() {
				if !defaultIgnore[entry.Name()] {
					stack = append(stack, full)
				}
			} else if entry.Type().IsRegular() {
				if len(extSet) == 0 || extSet[strings.ToLower(filepath.Ext(entry.Name()))] {
					results = append(results, full)
				}
			}
		}
	}
	sort.Strings(results)
	return results
}
